File: 116. Populating Next Right Pointers in Each Node.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
    if (root) {
        (function c(left, right) {
            if (left && right) {
                left.next = right;
                c(left.left, left.right)
                c(left.right, right.left)
                c(right.left, right.right)
            }
        })(root.left, root.right);
    }
    return root;
};

File: Golang/Tree/116.populating-next-right-pointers-in-each-node.go
package main

/**
 * Definition for a binary tree node.
 */
type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	if root != nil {
		stack := []*Node{root}
		for len(stack) > 0 {
			children := []*Node{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				if i < len(stack)-1 {
					node.Next = stack[i+1]
				}
				if node.Left != nil {
					children = append(children, node.Left)
				}
				if node.Right != nil {
					children = append(children, node.Right)
				}
			}
			stack = children
		}
	}
	return root
}
